contracts/IdentityRegistryBridger.sol

```diff
@@ -0,0 +1,328 @@
+/// @author Andrey Lebedev - <andrey.lebedev@valory.xyz>
+/// @author Mariapia Moscatiello - <mariapia.moscatiello@valory.xyz>
+contract IdentityRegistryBridger is ERC721TokenReceiver {
+    event OwnerUpdated(address indexed owner);
+    event ImplementationUpdated(address indexed implementation);
+    event AgentDecoupled(uint256 indexed serviceId, uint256 indexed agentId);
+    event AgentRegistered(uint256 indexed serviceId, uint256 indexed agentId, address serviceMultisig, string serviceTokenUri);
+    event AgentUpdated(uint256 indexed serviceId, uint256 indexed agentId, address serviceMultisig, string serviceTokenUri);
+
+    // Version number
+    string public constant VERSION = "0.1.0";
+    // Identity Registry Bridger proxy address slot
+    // keccak256("PROXY_IDENTITY_REGISTRY_BRIDGER") = "0x03684189c8fb7a536ac4dbd4b7ad063c37db21bcd0f9c51fe45a4eb16359c165"
+    bytes32 public constant PROXY_IDENTITY_REGISTRY_BRIDGER = 0x03684189c8fb7a536ac4dbd4b7ad063c37db21bcd0f9c51fe45a4eb16359c165;
+
+    // Identity Registry 8004 address
+    address public immutable identityRegistry;
+    // Service Manager address
+    address public immutable serviceManager;
+    // Service Registry address
+    address public immutable serviceRegistry;
+
+    // Owner address
+    address public owner;
+
+    // Reentrancy lock
+    uint256 internal _locked = 1;
+
+    // Mapping of service Id => 8004 agent Id
+    mapping(uint256 => uint256) public mapServiceIdAgentIds;
+
+    /// @dev IdentityRegistryBridger constructor.
+    /// @param _serviceManager Service Manager address.
+    constructor (address _identityRegistry, address _serviceManager) {
+        // Check for zero addresses
+        if (_identityRegistry == address(0) || _serviceManager == address(0)) {
+            revert ZeroAddress();
+        }
+
+        identityRegistry = _identityRegistry;
+        serviceManager = _serviceManager;
+        serviceRegistry = IServiceManager(_serviceManager).serviceRegistry();
+    }
+
+    /// @dev Registers 8004 agent Id corresponding to service Id.
+    /// @param serviceId Service Id.
+    /// @return agentId Corresponding 8004 agent Id.
+    function _register(uint256 serviceId) internal returns (uint256 agentId) {
+        // Get token URI
+        string memory tokenUri = IERC721(serviceRegistry).tokenURI(serviceId);
+
+        // Get actual service multisig
+        (,address multisig,,,,,) = IServiceRegistry(serviceRegistry).mapServices(serviceId);
+
+        // Check for zero address, although this must never happen
+        if (multisig == address(0)) {
+            revert ZeroAddress();
+        }
+
+        // TODO Work on this block to finalize
+        // Assemble OLAS specific metadata entry
+        IIdentityRegistry.MetadataEntry[] memory metadataEntries = new IIdentityRegistry.MetadataEntry[](1);
+        metadataEntries[0] = IIdentityRegistry.MetadataEntry({
+            key: "Ecosystem",
```

Comment:
Please check the spec if we use the key correctly

Also they told me in the chat that the spec sets a number of keys as MUST - check which ones